OGM-366 Neo4j remote integration

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/OGM-366

@gsmet I've appliedthe refactoring we discussed, I also had to rebase to the latest changes.

I've decided not to add the `QueryContext` for now and continue to use the `TupleContext`. I deal with the case where a session or an `EntityMetdataInfo` is null by throwing an exception if the dialect try to access some of the methods in the TupleContext when it shouldn't. I prefer this solution because it seems there are already other ways to guard against this and I don't want to update the signature in this PR  that it is already quite big.

gsmet commented on core/src/test/java/org/hibernate/ogm/backendtck/associations/manytoone/ManyToOneTest.java:
```diff
@@ -42,8 +42,8 @@ public void testUnidirectionalManyToOne() throws Exception {
+		assertThat( getNumberOfEntities( session ) ).isEqualTo( 3 );
+		assertThat( getNumberOfAssociations( session ) ).isEqualTo( expectedAssociations() );
 		transaction.commit();
 		assertThat( getNumberOfEntities( sessions ) ).isEqualTo( 3 );
```
Is there a reason why you didn't change all the calls and get rid of the SessionFactory based versions?

gsmet replied:
Apart from giving me less things to review, of course, which is a very valid reason ;)

DavideD replied:
> Apart from giving me less things to review, of course, which is a very valid reason ;)
 
That's the main reason, I prefer to keep the amount of changes to a minimum. For this cases in particular, we might have some tests that don't have a session and in general there is nothing wrong about using the factory. This change is about giving an additional option for those scenario where you need to make some assertions inside a specific transaction.

gsmet commented:
090fbd951a86bc485cb0683e9ba794ac8917b70b OGM-366 Create GridDialectType.NEO4J_REMOTE and skip tests

Typos in the commit comment:
suppose -> supposed
max indexer -> mass indexer

gsmet commented on core/src/test/java/org/hibernate/ogm/backendtck/associations/collection/manytomany/ManyToManyTest.java:
```diff
@@ -135,7 +135,7 @@ private void delete(Session session, Object... entities) {
 
 	private int expectedAssociationNumber() {
-		if ( TestHelper.getCurrentDialectType().equals( GridDialectType.NEO4J ) ) {
+		if ( TestHelper.getCurrentDialectType().equals( GridDialectType.NEO4J ) || TestHelper.getCurrentDialectType().equals( GridDialectType.NEO4J_REMOTE )) {
```
Minor nitpicking but we might as well be consistent in the way we test for this. See the change just above this one.

DavideD replied:
I'm not sure I understand what you are referring to  ...

gsmet replied:
See here:
https://github.com/hibernate/hibernate-ogm/pull/687/commits/090fbd951a86bc485cb0683e9ba794ac8917b70b#diff-0d9a048ba0f413c22966f840a46688a1L75
and here:
https://github.com/hibernate/hibernate-ogm/pull/687/commits/090fbd951a86bc485cb0683e9ba794ac8917b70b#diff-ca96314f78e37673af07f179b8c1f69eL138
or here:
https://github.com/hibernate/hibernate-ogm/pull/687/commits/090fbd951a86bc485cb0683e9ba794ac8917b70b#diff-c680f98d971b14f510ea306001b55d4aL120

Just a matter of consistency.

gsmet commented on core/src/test/java/org/hibernate/ogm/backendtck/associations/manytoone/ManyToOneTest.java:
```diff
@@ -117,7 +117,7 @@ public void canPersistAssociationToDetachedEntity() throws Exception {
 
 	private Long expectedAssociations() {
-		if ( TestHelper.getCurrentDialectType() == GridDialectType.NEO4J ) {
+		if ( TestHelper.getCurrentDialectType() == GridDialectType.NEO4J || TestHelper.getCurrentDialectType() == GridDialectType.NEO4J_REMOTE ) {
```
Same here.

gsmet commented:
51f3d99988e7c46f69e2f635c353973ee7707740 Add classes mapping the JSON reponse from the Neo4j server

Typos in the commit comment:
a a -> ?
realtive -> relative

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/json/impl/Statements.java:
```diff
@@ -0,0 +1,53 @@
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class Statements {
+
+	private List<Statement> statements = new ArrayList<Statement>( 5 );
```
Why 5?

DavideD replied:
Good question, I think it was a guesstimate, but I'll remove it.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/DatabaseIdentifier.java:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	/**
+	 * Returns the URI of the Neo4j server, e.g. "http:://localhost:5984".
```
s/::/:/

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/DatabaseIdentifier.java:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	/**
+	 * Returns the URI of the database, e.g. "http:://localhost:5984/mydb".
```
s/::/:/

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/DatabaseIdentifier.java:
```diff
@@ -0,0 +1,93 @@
+		this.userName = userName;
+		this.password = password;
+
+		this.serverUri = new URL( PROTOCOL, host, port, "" ).toURI();
```
Is there a reason why the protocol is hardcoded? I would have expected us to support HTTPS? That being said, we can fix it later.

DavideD replied:
No, no particular reason, this is a copy of an existing class.
 
I think it deserve a separate issue

gsmet replied:
I created a JIRA for this.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/Neo4jClient.java:
```diff
@@ -0,0 +1,120 @@
+	/**
+	 * Size of the client connection pool used by the RestEasy HTTP client
+	 */
+	private static final int CONNECTION_POOL_SIZE = 10;
```
Shouldn't we allow to configure this?

DavideD replied:
yes, I think it can be a follow-up commit or JIRA

gsmet replied:
I created a JIRA for this.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/Neo4jClient.java:
```diff
@@ -0,0 +1,120 @@
+	}
+
+	public Response authenticate(String username) {
+		return authenticationClient.authenticate( username );
```
Do I understand things correctly: authenticate doesn't do a stateful authentication, it just tests that the authentication is OK and the authentication is done using basic authentication for each call?

DavideD replied:
Yes, 
 it seems thought that it is an optional method and  I'm not using it anywhere. I think I'm going to remove it.

gsmet replied:
OK, maybe we could use it at startup to check that we can effectively connect to the Neo4j remote server?

DavideD replied:
ok, I'll check

gsmet replied:
If it's not possible, we might as well remove this method and the Rest interface.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/transaction/impl/Transaction.java:
```diff
@@ -0,0 +1,40 @@
+	}
+
+	public Long getId() {
+		return txId.longValue();
```
Either you want to return a primitive long or you don't need the longValue() call.

gsmet commented:
0425d7e816699997b8ac3a7a02552f2885ce48bf Add Neo4j client
iwll -> will
althought -> although

gsmet commented:
1e0fca3dd584699e73caf5e49eba5c4f5b763784 Extract reusable parts from the Neo4j embedded dialect

Typos in the commit comment:
extract -> extracts
part -> parts
tne -> the

gsmet commented on core/src/main/java/org/hibernate/ogm/util/impl/Log.java:
```diff
@@ -292,4 +292,7 @@
 	@Message(id = 86, value = "Transaction identifier not available")
 	HibernateException transactionIdIsNotAvailable();
+
+	@Message(id = 87, value = "The tuple context is not available, probably because we are dealing with more than a single entity")
```
I think "single entity **type**" would be better.

gsmet commented on core/src/main/java/org/hibernate/ogm/util/impl/TupleContextHelper.java:
```diff
@@ -6,29 +6,103 @@
+	}
+
+	/**
+	 * This {@link TupleContext} can be used for those use case where we don't necessarily have a tuple context but we
```
use case -> use cases

gsmet commented on core/src/main/java/org/hibernate/ogm/util/impl/TupleContextHelper.java:
```diff
@@ -6,29 +6,103 @@
+	 * This {@link TupleContext} can be used for those use case where we don't necessarily have a tuple context but we
+	 * still need a {@link TransactionContext}.
+	 */
+	private static class OnlyWithTransactionContext implements TupleContext {
```
I understand the point you exposed in your PR comment as the PR is already big enough. But I don't like this.

I think we should try to get rid of this once we got this PR merged. @gunnarmorling any opinion about this?

DavideD replied:
> I think we should try to get rid of this once we got this PR merged.

Sounds good

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/dialect/impl/AssociationQueries.java:
```diff
@@ -0,0 +1,604 @@
+ *
+ * @author Davide D'Alto
+ */
+public class AssociationQueries extends QueriesBase {
```
Shouldn't this be abstract?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/dialect/impl/EntityQueries.java:
```diff
@@ -0,0 +1,667 @@
+/**
+ * @author Davide D'Alto
+ */
+public class EntityQueries extends QueriesBase {
```
Shouldn't this be abstract?

gsmet commented on core/src/main/java/org/hibernate/ogm/datastore/impl/AvailableDatastoreProvider.java:
```diff
@@ -19,6 +19,7 @@
 	EHCACHE( "org.hibernate.ogm.datastore.ehcache.impl.EhcacheDatastoreProvider" ),
 	MONGODB( "org.hibernate.ogm.datastore.mongodb.impl.MongoDBDatastoreProvider" ),
 	FONGO( "org.hibernate.ogm.datastore.mongodb.impl.FongoDBDatastoreProvider" ),
+	NEO4J( "org.hibernate.ogm.datastore.neo4j.remote.impl.RemoteNeo4jDatastoreProvider" ),
```
Shouldn't it be called NEO4J_REMOTE?

DavideD replied:
At the moment, I'm using NEO4J for the remote and NEO4J_EMBEDDED for the embedded with the assumption that neo4j remote is the most common use case. NEO4J is much nicer to use.

gsmet replied:
Yeah well, my problem is that it's not consistent at all with the code where Neo4j* is for embedded (most of the time) and RemoteNeo4j* for remote.

I'll post a general comment on this at the end of my review so that we can deal with this.

gsmet commented on neo4j/pom.xml:
```diff
@@ -156,4 +156,25 @@
+        <profile>
+            <id>neo4j-remote</id>
+            <properties>
+                <datastoreProvider>neo4j</datastoreProvider>
```
I would have expected neo4j_remote here.

DavideD replied:
see previous comment

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/RemoteNeo4jDialect.java:
```diff
@@ -0,0 +1,559 @@
+ *
+ * @author Davide D'Alto &lt;davide@hibernate.org&gt;
+ */
+public class RemoteNeo4jDialect extends BaseNeo4jDialect {
```
I'm still a bit worried about this class. It duplicates quite a lot of things.

Have you tried to use generics on BaseNeo4jDialect with a couple of methods like newEntityQueries(),... , newTupleSnapshot() and so on?

Not sure it's a good idea but it might be worth a try if you haven't already done so.

DavideD replied:
I'll have a look.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/Neo4jProperties.java:
```diff
@@ -34,6 +34,31 @@
+	/**
+	 * Socket inactivity timeout in milliseconds.
+	 */
+	public static String SOCKET_TIMEOUT = "hibernate.ogm.neo4j.client.socketTimeout";
```
socket_timeout would be more consistent with the others?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/Neo4jProperties.java:
```diff
@@ -34,6 +34,31 @@
+	/**
+	 * Property for setting the password to connect with. Accepts {@code String}.
+	 */
+	public static String TTL_TIMEOUT = "hibernate.ogm.neo4j.client.socketTimeOut";
```
There's something wrong here: the comment is wrong and the property is the same as SOCKET_TIMEOUT. Looks like a leftover as it's not used anywhere. I think we can remove it.

DavideD replied:
I read the value in `Neo4jConfiguration` and then is set when the `Neo4jClient` is created (Neo4jClient#createRestClient())

gsmet replied:
Nope, not for this one. It's SOCKET_TIMEOUT that you read.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/Neo4jProperties.java:
```diff
@@ -34,6 +34,31 @@
+	/**
+	 * How long will we wait (in milliseconds) to get a connection?
+	 */
+	public static final String CONNECTION_CHECKOUT_TMEOUT = "hibernate.ogm.neo4j.client.connection_checkout_timeout";
```
TMEOUT -> TIMEOUT

gsmet commented on core/src/test/java/org/hibernate/ogm/test/util/impl/TupleContextHelperTest.java:
```diff
@@ -0,0 +1,90 @@
+import org.junit.rules.ExpectedException;
+
+/**
+ * Test the {@link TupleContextHelper} utility classe.
```
classe -> class
you're so French :)

DavideD replied:
In italian is "classe" :)

gsmet commented on core/src/test/java/org/hibernate/ogm/test/util/impl/TupleContextHelperTest.java:
```diff
@@ -0,0 +1,90 @@
+/**
+ * Test the {@link TupleContextHelper} utility classe.
+ * <p>
+ * An {@link EntityMetadataInformation} will be null when there is more than one entity is involved.
```
there is more than one entity is involved -> there is more than one entity type involved

gsmet commented on core/src/test/java/org/hibernate/ogm/test/util/impl/TupleContextHelperTest.java:
```diff
@@ -0,0 +1,90 @@
+
+	@Test
+	public void testTupleContextIsNull() throws Exception {
+		// This can happen when we start the mass-indexer without a session on more than one entity.
```
one entity -> one entity type

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+	}
+
+	/**
+	 * Create the sequence nodes setting the initial value if the node does not exists already.
```
exists -> exist

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+	/**
+	 * Query for creating SEQUENCE nodes.
+	 */
+	protected static final String SEQUENCE_CREATION_QUERY = "MERGE (n:" + NodeLabel.SEQUENCE.name() + " {" + SEQUENCE_NAME_PROPERTY + ": {sequenceName}} ) RETURN n";
```
Why is the creation query different from the embedded one? It doesn't take into account initialValue?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+	protected static final String SEQUENCE_CREATION_QUERY = "MERGE (n:" + NodeLabel.SEQUENCE.name() + " {" + SEQUENCE_NAME_PROPERTY + ": {sequenceName}} ) RETURN n";
+
+	protected static final String SEQUENCE_LOCK_QUERY = "MATCH (n:" + NodeLabel.SEQUENCE.name() + ") WHERE n." + SEQUENCE_NAME_PROPERTY + " = {sequenceName} "
+			+ " SET n.__locked = true " + " RETURN n";
```
Is there a reason why the locks are managed differently between embedded and remote?

DavideD replied:
Cypher does not provide a way to acquire a lock on a node. The lock property only purpose is to make sure that a lock is acquired before updating the sequence.
In embedded mode, I can use a nice API and I can acquire a lock on a single node easily.
Therefore I prefer to use a different approach in the two cases.

gsmet replied:
OK, I must admit that I'm a bit worried about the level of duplications we have here. If we consider the embedded case to be a not so common case, maybe we should avoid as much duplication as possible even if it's a little less effective/nice. Apart from the execution of the queries/extraction of the result, I would have expected both classes to be essentially the same.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+	 * @param initialValue the initial value of the given generator
+	 * @return the next value in a sequence
+	 */
+	public Number nextValue(IdSourceKey idSourceKey, int increment, int initialValue) {
```
The parameters increment and initialValue are interverted between the embedded and the remote versions. Doesn't look like a good idea. Maybe having an abstract method in the parent class would be handy to avoid this?

DavideD replied:
Mmmh... are you talking about RemoteNeo4jSequenceGenerator#nextvalue and Neo4jSequenceGenerator#nextValue? I don't see a difference...

But I see a difference between nextValue and the private methods that follow

gsmet replied:
Oh, yes sorry. It's effectively between nextValue and all the private methods generating queries.

Btw, my comment on the abstract method still stands as the return type is not the same and it would probably be better for these methods to be consistent. (This if we cannot find a way to avoid the code duplication)

DavideD replied:
> my comment on the abstract method still stands a ...

no problem, I will add it

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+	 * Generate the next value in a sequence for a given {@link IdSourceKey}.
+	 *
+	 * @param idSourceKey identifies the generator
+	 * @param increment the difference between to consecutive values in the sequence
```
to -> two

gsmet replied:
Btw, same typo in the embedded one, probably a preexisting one.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+				throw new HibernateException( "Sequence missing: " + sequenceName );
+			}
+		}
+		return nextValue.longValue() - increment;
```
This one looks weird, why would you substract the increment from the nextValue?

DavideD replied:
The only way I found to make it work in a multi-threaded environment is to first increment the value and then read it . Our API allows for an initial value and to make sure that I'm actually reading the correct one the first time I need to decrement the value I obtain from the db.

gsmet replied:
OK got it, thanks. Also removed a comment above asking why the creation of the sequence was different and not taking into account the initialValue.

Might be worth a comment before `return nextValue.longValue() - increment;` though!

DavideD replied:
Sure, I will add it

On 6 Jun 2016 17:37, "Guillaume Smet" <notifications@github.com> wrote:
>
> In
neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
>
> > + * @param initialValue the initial value of the given generator
> > + * @return the next value in a sequence
> > + */
> > + public Number nextValue(IdSourceKey idSourceKey, int increment, int
initialValue) {
> > + String sequenceName = sequenceName( idSourceKey );
> > + Statements statements = updateNextValueQuery( idSourceKey,
initialValue, increment );
> > + StatementsResponse statementsResponse =
neo4jDb.executeQueriesInNewTransaction( statements );
> > + List<StatementResult> results = statementsResponse.getResults(
> );
> > + Integer nextValue = (Integer) results.get( 1 ).getData().get( 0
).getRow().get( 0 );
> > + // sequence nodes are expected to have been created up-front
> > + if ( idSourceKey.getMetadata().getType() == IdSourceType.SEQUENCE ) {
> > + if ( nextValue == null ) {
> > + throw new HibernateException( "Sequence missing: " + sequenceName );
> > + }
> > + }
> > + return nextValue.longValue() - increment;
>
> OK got it, thanks. Also removed a comment above asking why the creation
of the sequence was different and not taking into account the initialValue.
>
> Might be worth a comment before return nextValue.longValue() - increment;
though!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jEntityQueries.java:
```diff
@@ -0,0 +1,334 @@
+		String query = getFindEntityQuery();
+		List<StatementResult> results = executeQuery( executionEngine, transactionId, query, params );
+		if ( results != null ) {
+			Row row = results.get( 0 ).getData().get( 0 );
```
I'm a bit surprised by the structure of the results which look overly complex.

DavideD replied:
In a single request you can execute multiple queries and each query result can be returned as rows of a  table or as graph elements. All this make things a bit complicated

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/RemoteNeo4jSchemaDefiner.java:
```diff
@@ -0,0 +1,119 @@
+		List<StatementResult> results = response.getResults();
+		// TODO: Check response for errors
+
+		// We create the sequencese in a separate transaction because
```
sequencese -> sequences

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/RemoteNeo4jSchemaDefiner.java:
```diff
@@ -0,0 +1,119 @@
+
+		StatementsResponse response = provider.getDatabase().executeQueriesInNewTransaction( constraintStatements );
+		List<StatementResult> results = response.getResults();
+		// TODO: Check response for errors
```
Maybe open a JIRA if you don't plan to fix them in this PR?

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/impl/RemoteNeo4jSchemaDefiner.java:
```diff
@@ -0,0 +1,119 @@
+		provider.getSequenceGenerator().createSequences( sequenceStatements, sequences );
+		response = provider.getDatabase().executeQueriesInNewTransaction( sequenceStatements );
+		results = response.getResults();
+		// TODO: Check response for errors
```
Maybe open a JIRA if you don't plan to fix them in this PR?

DavideD replied:
Yes

DavideD replied:
By the way, no need to add a coment for evey TODO (unless is something different), I will open a JIRA for them

gsmet replied:
There are only 2 of them!

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/transaction/impl/RemoteJtaTransactionCoordinator.java:
```diff
@@ -0,0 +1,153 @@
+		}
+	}
+
+	private class RemoteTransactionDriver extends ForwardingTransactionDriver implements IdentifiableDriver {
```
Is there a reason why you don't use the same pattern as for Neo4jJtaTransactionCoordinator? Neo4jJtaTransactionCoordinator just relies on Neo4jSynchronization.

DavideD replied:
Yes, in the case of remote I need the transaction id to execute Rest API calls from the dialect

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/transaction/impl/RemoteResourceLocalTransactionCoordinator.java:
```diff
@@ -0,0 +1,339 @@
+/**
+ * An implementation of TransactionCoordinator based on managing a transaction through a Neo4j Connection.
+ * <p>
+ * This is inspired by {@code org.hibernate.resource.transaction.backend.jdbc.internal.JdbcResourceLocalTransactionCoordinatorImp}.
```
Imp -> Impl in the class name.

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/transaction/impl/RemoteResourceLocalTransactionCoordinator.java:
```diff
@@ -0,0 +1,339 @@
+ *
+ * @author Davide D'Alto
+ */
+public class RemoteResourceLocalTransactionCoordinator implements TransactionCoordinator {
```
Maybe we could avoid a good part of the duplication between this and Neo4jResourceLocalTransactionCoordinator?

gsmet commented:
OK, this is by far my largest review so far.

Overall, nice work, it does the job and it was quite easy to review the split commits.

2 general remarks:
- I wonder if we could avoid some of the duplication we have between the 2 dialects. I put some comments where I was worried about this. If it's not feasible, we can keep it like that but I think we should try to reduce the duplication as much as possible - as you already did it in several places;
- we are not consistent about the naming:
  - sometimes, it's Neo4j for embedded and Neo4jRemote for remote
  - sometimes it's the opposite (not in the code but in the configuration)
  - a few classes have the Neo4j for embedded but are only called Remote for remote (the transaction coordinator stuff)
  - for the SchemaDefiner, we have EmbeddedNeo4jSchemaDefiner for embedded
  - sometimes, the base implementation is called BaseNeo4j* and sometimes it doesn't have a prefix at all such as EntityQueries and AssociationQueries.


-> I think we should try to be as consistent as possible and I would propose the following naming scheme:
- EmbeddedNeo4j and RemoteNeo4j prefix for all the classes specific to a a dialect
- BaseNeo4j prefix for the abstract classes which are then in 2 flavors
- Neo4j prefix if the class is common to both without any difference (I don't propose to rename the existing class, just to be consistent for the ones which are added in 2 flavors)

gsmet commented on neo4j/src/main/java/org/hibernate/ogm/datastore/neo4j/remote/dialect/impl/RemoteNeo4jSequenceGenerator.java:
```diff
@@ -0,0 +1,253 @@
+		Statements statements = updateNextValueQuery( idSourceKey, initialValue, increment );
+		StatementsResponse statementsResponse = neo4jDb.executeQueriesInNewTransaction( statements );
+		List<StatementResult> results = statementsResponse.getResults();
+		Integer nextValue = (Integer) results.get( 1 ).getData().get( 0 ).getRow().get( 0 );
```
This looks a bit like black magic :)

DavideD commented:
I've applied all the remarks:
* Fix typos (Hopefully I haven't added any new one)
* Added credential authentication when the client is created
* Added more comments around
* Remove TODO: Throw exception in case of error during sequences or constraints creation
* Change AvailableDatastoreProvider.NEO4J to AvailableDatastoreProvider.NEO4J_REMOTE
* Add the prefix RemoteNeo4j to cthe classes related to the remote Neo4j integration.

As for renaming the other classes I will create a separate JIRA for it since I was also thinking about having different packages for remote and embedded and this will have an impact on many files.

I will create a separate issue for adding back the QueryContext as well

DavideD commented:
> As for renaming the other classes I will create a separate JIRA for it since I was also thinking about having different packages for remote and embedded and this will have an impact on many files.

It might also be cause of discussion, so I would like to see this PR integrated first

DavideD commented:
> I wonder if we could avoid some of the duplication we have between the 2 dialects. I put some comments where I was worried about this. If it's not feasible, we can keep it like that but I think we should try to reduce the duplication as much as possible - as you already did it in several places;

I think as a first PR this is clean enough really, we can continue refactoring while working on other things. I think it will be easier to do it once I've renamed the existing classes and created some additional packages.

gsmet commented:
Superseeded by https://github.com/hibernate/hibernate-ogm/pull/690 and merged!
